File: 프로그래머스/1/42889. 실패율/실패율.py
# from collections import Counter
# def solution(N, stages):
#     stages.sort()
#     users = len(stages)
#     result = dict()
#     ans = []
    
#     for i in range(1,N+1):
#         counts = stages.count(i)
        
#         fail = counts / users
        
#         users = users - counts
        
#         result[i] = fail
        
#     sort_result = sorted(result.items() , key = lambda x : x[1], reverse=True)
    
#     for a,b in sort_result:
#         ans.append(a)
        
#     return ans
        
        
from collections import Counter

def solution(N, stages):
    stage_counts = Counter(stages)
    users = len(stages)
    result = dict()

    for i in range(1, N+1):
        count = stage_counts[i]  # O(1)
        if users == 0:
            result[i] = 0
        else:
            result[i] = count / users

        users -= count

    return sorted(result, key=lambda x: result[x], reverse=True)

    
    

File: 프로그래머스/1/42889. 실패율/실패율.js
function solution(N, stages) {
    const stage_counts = {}
    let users = stages.length
    let result = {}
    
    for(const v of stages){
        stage_counts[v] = (stage_counts[v] || 0) + 1;
    }
    
    for(let i=1; i<N+1; i++){
        let counts = stage_counts[i]
        if(counts == undefined){
            counts = 0
        }
        
        if (users === 0){
            result[i] = 0
        }
        else{
            result[i] = counts / users
        }
        
        users = users - counts
    }
    
    const ans = Object.entries(result).sort((a,b) => b[1] - a[1]).map(([key]) => +key)
    
    return ans
            
}